src/util/logger.rs

```diff
@@ -0,0 +1,193 @@
+// Pruned copy of crate rust log, without global logger
+// https://github.com/rust-lang-nursery/log #7a60286
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// There is currently 2 ways to filter log messages. First one, by using compilation flag, e.g "max_level_off".
+/// The second one, client-side by implementing check against Record Level field, e.g TestLogger in test_utils.
+/// Each module may have its own Logger or share one.
+
+use std::cmp;
+use std::fmt;
+
+static LOG_LEVEL_NAMES: [&'static str; 6] = ["OFF", "ERROR", "WARN", "INFO", "DEBUG", "TRACE"];
+
+/// An enum representing the available verbosity levels of the logger.
+#[derive(Copy, Clone, Eq, Debug, Hash)]
+pub enum Level {
+	///Designates logger being silent
+	Off,
+	/// Designates very serious errors
+	Error,
+	/// Designates hazardous situations
+	Warn,
+	/// Designates useful information
+	Info,
+	/// Designates lower priority information
+	Debug,
+	/// Designates very low priority, often extremely verbose, information
+	Trace,
+}
+
+impl PartialEq for Level {
+	#[inline]
+	fn eq(&self, other: &Level) -> bool {
+		*self as usize == *other as usize
+	}
+}
+
+impl PartialOrd for Level {
+	#[inline]
+	fn partial_cmp(&self, other: &Level) -> Option<cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+
+	#[inline]
+	fn lt(&self, other: &Level) -> bool {
+		(*self as usize) < *other as usize
+	}
+
+	#[inline]
+	fn le(&self, other: &Level) -> bool {
+		*self as usize <= *other as usize
+	}
+
+	#[inline]
+	fn gt(&self, other: &Level) -> bool {
+		*self as usize > *other as usize
+	}
+
+	#[inline]
+	fn ge(&self, other: &Level) -> bool {
+		*self as usize >= *other as usize
+	}
+}
+
+impl Ord for Level {
+	#[inline]
+	fn cmp(&self, other: &Level) -> cmp::Ordering {
+		(*self as usize).cmp(&(*other as usize))
+	}
+}
+
+impl fmt::Display for Level {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		fmt.pad(LOG_LEVEL_NAMES[*self as usize])
+	}
+}
+
+impl Level {
+	/// Returns the most verbose logging level.
+	#[inline]
+	pub fn max() -> Level {
+		Level::Trace
+	}
+}
+
+/// A Record, unit of logging output with Metadata to enable filtering
+/// Module_path, file, line to inform on log's source
+#[derive(Clone,Debug)]
+pub struct Record<'a> {
+	level: Level,
+	args: fmt::Arguments<'a>,
+	module_path: &'a str,
+	file: &'a str,
+	line: u32,
+}
+
+impl<'a> Record<'a> {
+	/// Returns a new Record.
+	#[inline]
+	pub fn new(level: Level, args: fmt::Arguments<'a>, module_path: &'a str, file: &'a str, line: u32) -> Record<'a> {
+		Record {
+			level,
+			args,
+			module_path,
+			file,
+			line
+		}
+	}
+
+	/// The message body.
+	#[inline]
+	pub fn args(&self) -> &fmt::Arguments<'a> {
```

Comment:
Why not just make the vars in Record pub (with docs) and not worry about accessors?